cmd/repo-updater/repos/types.go

```diff
@@ -922,3 +926,89 @@ func (es ExternalServices) With(opts ...func(*ExternalService)) ExternalServices
 	clone.Apply(opts...)
 	return clone
 }
+
+type RateLimiterRegistry struct {
+	mu sync.Mutex
+	// Rate limiter per external service
+	rateLimiters map[int64]*rate.Limiter
+}
+
+func NewRateLimiterRegistry(ctx context.Context, store Store) (*RateLimiterRegistry, error) {
+	svcs, err := store.ListExternalServices(ctx, StoreListExternalServicesArgs{})
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching external services")
+	}
+
+	r := &RateLimiterRegistry{
+		rateLimiters: make(map[int64]*rate.Limiter),
+	}
+
+	for _, svc := range svcs {
+		err = r.updateRateLimiter(svc)
+		if err != nil {
+			// Errors here are not fatal
+			log15.Warn("Updating rate limiter", "kind", svc.Kind, "err", err)
+		}
+	}
+
+	return r, nil
+}
+
+// GetRateLimiter fetches the rate limiter associated with the given external service. If none has been
+// configured an infinite limiter is returned.
+func (r *RateLimiterRegistry) GetRateLimiter(externalServiceID int64) *rate.Limiter {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	l := r.rateLimiters[externalServiceID]
+	if l == nil {
+		l = rate.NewLimiter(rate.Inf, 100)
+		r.rateLimiters[externalServiceID] = l
+	}
+	return l
+}
+
+// HandleExternalServiceSync will update the rate limiter associated with the supplied external service
+// so that it's settings match the service config.
+func (r *RateLimiterRegistry) HandleExternalServiceSync(apiService api.ExternalService) error {
+	svc := &ExternalService{
+		ID:          apiService.ID,
+		Kind:        apiService.Kind,
+		DisplayName: apiService.DisplayName,
+		Config:      apiService.Config,
+		CreatedAt:   apiService.CreatedAt,
+		UpdatedAt:   apiService.UpdatedAt,
+		DeletedAt:   time.Time{},
+	}
+	if apiService.DeletedAt != nil {
+		svc.DeletedAt = *apiService.DeletedAt
+	}
+	return r.updateRateLimiter(svc)
+}
+
+func (r *RateLimiterRegistry) updateRateLimiter(svc *ExternalService) error {
+	config, err := svc.Configuration()
+	if err != nil {
+		return errors.Wrap(err, "getting external service configuration")
+	}
+
+	var limit rate.Limit
+	switch c := config.(type) {
+	case *schema.GitLabConnection:
+		// 10/s is the default enforced by GitLab on their end
+		limit = rate.Limit(10)
+		if c != nil && c.RateLimit != nil {
+			if c.RateLimit.Enabled {
+				limit = rate.Limit(c.RateLimit.RequestsPerHour / 3600)
+			} else {
+				limit = rate.Inf
+			}
+		}
```

Comment:
Since GitLab has the hard limit, we probably don't want to have Inf if it's not configured (i.e. remain the default):
```suggestion
		if c != nil && c.RateLimit != nil && c.RateLimit.Enabled {
			limit = rate.Limit(c.RateLimit.RequestsPerHour / 3600)
		}
```

Reply:
We fall back to unlimited only when the admin has disabled rate limiting.  So this is the case where they've added rate limit config but set Enabled to false.

Reply:
Ah you're right, sorry! Probably worth putting in the comments 😛